rfd/0207-vnet-ssh.md

```diff
@@ -0,0 +1,398 @@
+
+Alice prefers to develop and run her application in Visual Studio Code and she
+knows it supports remote development so she decides to try connecting to
+`devbox` in VSCode from her laptop.
+She has Microsoft's official "Remote - SSH" extension installed in VSCode so she
+goes to the command palette and selects "Remote-SSH: Connect to Host..." and
+types in `ssh alice@devbox.ssh.teleport.example.com`.
+VSCode successfully connects to the host and Alice is able to use all the normal
+development workflows she is used to.
+
+#### User story - Per-session MFA
+
+Later, Alice's company enables per-session MFA for SSH connections in the
+cluster.
+The next time Alice tries to connect to `devbox` in VSCode, Connect steals focus
+and prompts her to tap her security key to allow access to the server.
+After she taps the security key, the connection succeeds.
+
+VNet caches the MFA-verified SSH certificate for 1 minute so that VSCode can
+start multiple connections to the host without additional prompts.
+Alice doesn't get another MFA prompt until her Teleport session expires or she
+closes and reopens the remote connection in VSCode.
+
+#### User story - Failed authentication
+
+Alice tries to connect to `devbox` with the SSH username `root` by running
+`ssh root@devbox.ssh.teleport.example.com`.
+However, her Teleport roles do not allow her to SSH as `root`.
+She gets an error message from `ssh` on the CLI reading
+`ERROR: access denied to root connecting to devbox`.
+
+### Implementation
+
+The purpose of VNet SSH is to simplify Teleport SSH connections with third-party
+SSH clients as much as possible.
+The goal is zero manual configuration steps for the user.
+In order to support per-session MFA in this scenario, the third-party SSH
+client cannot be required to use the Teleport user SSH certificate to terminate
+the SSH connection with the target SSH server, because it simply does not have a
+way to get the MFA-verified session certificate.
+
+This means that the VNet process must be responsible for dialing the target SSH
+server with the Teleport user certificate.
+This is similar to the way VNet already handles TCP app access, it:
+
+1. terminates the incoming TCP connection
+1. gets a user certificate for the target app
+1. TLS dials to the proxy address with the TLS client certificate
+1. forwards the TCP stream between the incoming local connection and the dialed
+   connection to the target app
+
+For VNet SSH we will:
+
+1. terminate the incoming SSH connection
+1. use the user's SSH certificate or get a per-session MFA certificate if
+   necessary
+1. SSH dial to the target SSH server with the user SSH certificate
+1. forward the full SSH connection between the incoming local connection and the
+   dialed connection to the target SSH server
+
+Forwarding a full SSH connection is slightly more complicated than forwarding a
+simple TCP connection, we have to handle all SSH channels, global requests,
+channel requests, etc.
```

Comment:
We should probably add more tests or test plan items for things like X11 and TCP forwarding through VNet SSH.